external/esp_idf_port/esp32/spi_ram/spiram.c

```diff
@@ -0,0 +1,270 @@
+//#include "freertos/FreeRTOS.h"
+//#include "freertos/xtensa_api.h"
+#include "chip/esp32_soc.h"
+#include "esp_heap_caps.h"
+#include "soc/soc_memory_layout.h"
+#include "chip/esp32_dport.h"
+#include "rom/cache.h"
+#include <tinyara/mm/heap_regioninfo.h>
+
+#ifndef CONFIG_SMP
+#define PSRAM_MODE PSRAM_VADDR_MODE_NORMAL
+#else
+#if CONFIG_MEMMAP_SPIRAM_CACHE_EVENODD
+#define PSRAM_MODE PSRAM_VADDR_MODE_EVENODD
+#else
+#define PSRAM_MODE PSRAM_VADDR_MODE_LOWHIGH
+#endif
+#endif
+
+#ifdef CONFIG_SPIRAM_SUPPORT
+
+//static const char* TAG = "spiram";
+
+#if CONFIG_SPIRAM_SPEED_40M && CONFIG_ESPTOOLPY_FLASHFREQ_40M
+#define PSRAM_SPEED PSRAM_CACHE_F40M_S40M
+#elif CONFIG_SPIRAM_SPEED_40M && CONFIG_ESPTOOLPY_FLASHFREQ_80M
+#define PSRAM_SPEED PSRAM_CACHE_F80M_S40M
+#elif CONFIG_SPIRAM_SPEED_80M && CONFIG_ESPTOOLPY_FLASHFREQ_80M
+#define PSRAM_SPEED PSRAM_CACHE_F80M_S80M
+#else
+#error "FLASH speed can only be equal to or higher than SRAM speed while SRAM is enabled!"
+#endif
+
+//manually redefined this size config
+#define CONFIG_SPIRAM_SIZE regionx_size[1]
+
+static bool spiram_inited=false;
+
+/*
+ Simple RAM test. Writes a word every 32 bytes. Takes about a second to complete for 4MiB. Returns
+ true when RAM seems OK, false when test fails. WARNING: Do not run this before the 2nd cpu has been
+ initialized (in a two-core system) or after the heap allocator has taken ownership of the memory.
+*/
+bool esp_spiram_test()
+{
+	volatile int *spiram = (volatile int *)SOC_EXTRAM_DATA_LOW;
+	size_t p;
+	size_t s = CONFIG_SPIRAM_SIZE;
+	int errct = 0;
+	int initial_err = -1;
+	for (p = 0; p < (s / sizeof(int)); p += 8) {
+		spiram[p] = p ^ 0xAAAAAAAA;
+	}
+	for (p = 0; p < (s / sizeof(int)); p += 8) {
+		if (spiram[p] != (p ^ 0xAAAAAAAA)) {
+			errct++;
+			if (errct == 1) {
+				initial_err = p * 4;
+			}
+		}
+	}
+	printf("read spiram over\n");
+	if (errct) {
+		printf("SPI SRAM test failed\n");
+		//ESP_EARLY_LOGE(TAG, "SPI SRAM memory test fail. %d/%d writes failed, first @ %X\n", errct, s/32, initial_err+SOC_EXTRAM_DATA_LOW);
+		return false;
+	} else {
+		printf("SPI SRAM test OK\n");
+		//ESP_EARLY_LOGI(TAG, "SPI SRAM memory test OK");
+		return true;
+	}
+}
+
+void IRAM_ATTR esp_spiram_init_cache()
+{
+	//Enable external RAM in MMU
+	cache_sram_mmu_set(0, 0, SOC_EXTRAM_DATA_LOW, 0, 32, 128);
+	//Flush and enable icache for APP CPU
+#ifdef CONFIG_SMP
+	DPORT_CLEAR_PERI_REG_MASK(DPORT_APP_CACHE_CTRL1_REG, DPORT_APP_CACHE_MASK_DRAM1);
+	cache_sram_mmu_set(1, 0, SOC_EXTRAM_DATA_LOW, 0, 32, 128);
+#endif
+}
+
+esp_err_t esp_spiram_init()
+{
+	esp_err_t r;
+	r = psram_enable(PSRAM_SPEED, PSRAM_MODE);
+	if (r != ESP_OK) {
+#if CONFIG_SPIRAM_IGNORE_NOTFOUND
+		ESP_EARLY_LOGE(TAG, "SPI RAM enabled but initialization failed. Bailing out.");
+#endif
+		ets_printf("SPI RAM enabled but initialization failed. Bailing out.");
+		return r;
+	}
+	//ESP_EARLY_LOGI(TAG, "SPI RAM mode: %s", PSRAM_SPEED == PSRAM_CACHE_F40M_S40M ? "flash 40m sram 40m" : \
```

Comment:
this print code is commented as esp log module is not ported at that time. 
now , esp log is ported, if merged, pls think revert it.

Reply:
reverted the esp log print, thanks